inox2d/src/render.rs

```diff
@@ -1,378 +1,386 @@
-		for id in self.nodes.root.descendants(&self.nodes.arena).skip(1) {
-			let node_index = &self.nodes.arena[id];
-			let node = node_index.get();
-
-			if node.lock_to_root {
-				let node_render_ctx = node_rctxs.get_mut(&node.uuid).unwrap();
-				node_render_ctx.trans = root_trans * node_render_ctx.trans_offset.to_matrix();
-			} else {
-				let parent = &self.nodes.arena[node_index.parent().unwrap()].get();
-				let parent_trans = node_rctxs.get(&parent.uuid).unwrap().trans;
-
-				let node_render_ctx = node_rctxs.get_mut(&node.uuid).unwrap();
-				node_render_ctx.trans = parent_trans * node_render_ctx.trans_offset.to_matrix();
-			}
-		}
+	/// Memory layout: `[[x, y], [x, y], ...]`
+	pub fn get_raw_verts(&self) -> &[f32] {
+		VertexBuffers::vec_vec2_as_vec_f32(&self.vertex_buffers.verts)
+	}
+	/// Memory layout: `[[x, y], [x, y], ...]`
+	pub fn get_raw_uvs(&self) -> &[f32] {
+		VertexBuffers::vec_vec2_as_vec_f32(&self.vertex_buffers.uvs)
+	}
+	/// Memory layout: `[[i0, i1, i2], [i0, i1, i2], ...]`
+	pub fn get_raw_indices(&self) -> &[u16] {
+		self.vertex_buffers.indices.as_slice()
+	}
+	/// Memory layout: `[[dx, dy], [dx, dy], ...]`
+	pub fn get_raw_deforms(&self) -> &[f32] {
+		VertexBuffers::vec_vec2_as_vec_f32(&self.vertex_buffers.deforms)
 	}
 }
 
+/// Same as the reference Inochi2D implementation, Inox2D also aims for a "bring your own rendering backend" design.
+/// A custom backend shall implement this trait.
+///
+/// It is perfectly fine that the trait implementation does not contain everything needed to display a puppet as:
+/// - The renderer may not be directly rendering to the screen for flexibility.
+/// - The renderer may want platform-specific optimizations, e.g. batching, and the provided implementation is merely for collecting puppet info.
+/// - The renderer may be a debug/just-for-fun renderer intercepting draw calls for other purposes.
+///
+/// Either way, the point is Inox2D will implement a `draw()` method for any `impl InoxRenderer`, dispatching calls based on puppet structure according to Inochi2D standard.
 pub trait InoxRenderer
 where
 	Self: Sized,
 {
+	type NewParams;
 	type Error;
 
-	/// For any model-specific setup, e.g. creating buffers with specific sizes.
+	/// Create a renderer for one model.
 	///
-	/// After this step, the provided model should be renderable.
-	fn prepare(&mut self, model: &Model) -> Result<(), Self::Error>;
-
-	/// Resize the renderer's viewport.
-	fn resize(&mut self, w: u32, h: u32);
-
-	/// Clear the canvas.
-	fn clear(&self);
-
-	/// Initiate one render pass.
-	fn on_begin_scene(&self);
-	/// The render pass.
-	///
-	/// Logical error if this puppet is not from the latest prepared model.
-	fn render(&self, puppet: &Puppet);
-	/// Finish one render pass.
-	fn on_end_scene(&self);
-	/// Actually make results visible, e.g. on a screen/texture.
-	fn draw_scene(&self);
+	/// Ref impl: Upload textures.
+	fn new(params: Self::NewParams, model: &Model) -> Result<Self, Self::Error>;
```

Comment:
This function is not used polymorphically inside the library, so I think it's better to not specify any need for it. It'll remove the need for the `NewParams` associated type and make the `new` function of every renderer more flexible.